spi/imx6ull-qspi/qspi.c

```diff
@@ -0,0 +1,326 @@
+#define QSPI_SFA2AD (qspi_common.base + 0x184 / sizeof(uint32_t))
+#define QSPI_SFB1AD (qspi_common.base + 0x188 / sizeof(uint32_t))
+#define QSPI_SFB2AD (qspi_common.base + 0x18C / sizeof(uint32_t))
+
+#define WATERMARK 16
+
+#define NUM_LUT_SEQ 16
+
+typedef struct {
+	int mux;
+	char val;
+	char sion;
+} qspi_pctl_mux_t;
+
+typedef struct {
+	int pad;
+	char speed;
+	char dse;
+	char sre;
+} qspi_pctl_pad_t;
+
+qspi_pctl_mux_t qspi_pctl_mux[2][8] = {
+	{ { pctl_mux_nand_d7, 2, 0 }, { pctl_mux_nand_ale, 2, 0 }, { pctl_mux_nand_wp, 2, 0 }, { pctl_mux_nand_rdy, 2, 0 }, { pctl_mux_nand_ce0, 2, 0 }, { pctl_mux_nand_ce1, 2, 0 }, { pctl_mux_nand_cle, 2, 0 }, { pctl_mux_nand_dqs, 2, 0 } },
+	{ { pctl_mux_nand_re, 2, 0 }, { pctl_mux_nand_we, 2, 0 }, { pctl_mux_nand_d0, 2, 0 }, { pctl_mux_nand_d1, 2, 0 }, { pctl_mux_nand_d2, 2, 0 }, { pctl_mux_nand_d3, 2, 0 }, { pctl_mux_nand_d4, 2, 0 }, { pctl_mux_nand_d5, 2, 0 } },
+};
+
+qspi_pctl_pad_t qspi_pctl_pad[2][8] = {
+	{ { pctl_pad_nand_d7, 1, 4, 0 }, { pctl_pad_nand_ale, 1, 4, 0 }, { pctl_pad_nand_wp, 1, 4, 0 }, { pctl_pad_nand_rdy, 1, 4, 0 }, { pctl_pad_nand_ce0, 1, 4, 0 }, { pctl_pad_nand_ce1, 1, 4, 0 }, { pctl_pad_nand_cle, 1, 4, 0 }, { pctl_pad_nand_dqs, 1, 4, 0 } },
+	{ { pctl_pad_nand_re, 1, 4, 0 }, { pctl_pad_nand_we, 1, 4, 0 }, { pctl_pad_nand_d0, 1, 4, 0 }, { pctl_pad_nand_d1, 1, 4, 0 }, { pctl_pad_nand_d2, 1, 4, 0 }, { pctl_pad_nand_d3, 1, 4, 0 }, { pctl_pad_nand_d4, 1, 4, 0 }, { pctl_pad_nand_d5, 1, 4, 0 } },
+};
+
+
+struct {
+	volatile uint32_t *base;
+	uint32_t flash_base_addr[2];
+	bool init;
+} qspi_common = { .base = NULL, .init = false, .flash_base_addr = { QSPI_MMAP_BASE, QSPI_MMAP_BASE + QSPI_FLASH_A1_SIZE + QSPI_FLASH_A2_SIZE } };
+
+// TODO ustawić ile ohmów w rejestrze na flashu.
+// TODO find clock frequency - ~50Mhz
+
+static void set_mux(qspi_dev_t dev_no)
+{
+	platformctl_t ctl;
+	int i;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_iomux;
+
+	for (i = 0; i < 8; i++) {
+		ctl.iomux.mux = qspi_pctl_mux[dev_no][i].mux;
+		ctl.iomux.mode = qspi_pctl_mux[dev_no][i].val;
+		ctl.iomux.sion = qspi_pctl_mux[dev_no][i].sion;
+		platformctl(&ctl);
+	}
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_iopad;
+
+	for (i = 0; i < 8; i++) {
+		ctl.iopad.pad = qspi_pctl_pad[dev_no][i].pad;
+		ctl.iopad.speed = qspi_pctl_pad[dev_no][i].speed;
+		ctl.iopad.dse = qspi_pctl_pad[dev_no][i].dse;
+		ctl.iopad.sre = qspi_pctl_pad[dev_no][i].sre;
+		platformctl(&ctl);
+	}
+}
+
+static void set_clk()
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_devclock;
+	ctl.devclock.dev = pctl_clk_qspi;
+	ctl.devclock.state = 0x03;
+
+	platformctl(&ctl);
+}
+
+int qspi_setLutSeq(const lut_seq_t *lut, unsigned int lut_seq)
+{
+	if (lut_seq >= NUM_LUT_SEQ) {
+		return -1;
+	}
+	if (*QSPI_LCKCR_ADDR & QSPI_LCKCR_LOCK) {
+		*QSPI_LUT_KEY_ADDR = QSPI_LUT_KEY;
+		*QSPI_LCKCR_ADDR = (*QSPI_LCKCR_ADDR & ~(QSPI_LCKCR_UNLOCK | QSPI_LCKCR_LOCK)) | QSPI_LCKCR_UNLOCK;
+	}
+
+	for (int i = 0; i < 4; i++) {
```

Comment:
For now we still use K&R/C89 variable declaration style, i.e. on the start of the function only
```suggestion
	int i;
	
	if (lut_seq >= NUM_LUT_SEQ) {
		return -1;
	}
	if (*QSPI_LCKCR_ADDR & QSPI_LCKCR_LOCK) {
		*QSPI_LUT_KEY_ADDR = QSPI_LUT_KEY;
		*QSPI_LCKCR_ADDR = (*QSPI_LCKCR_ADDR & ~(QSPI_LCKCR_UNLOCK | QSPI_LCKCR_LOCK)) | QSPI_LCKCR_UNLOCK;
	}

	for (i = 0; i < 4; i++) {
```